cmd/roachtest: adapt disk-stall detection roachtest

jbowens opened the pull request:
Move the existing disk-stall/* roachtests under disk-stall/fuse/* (for the FUSE
filesystem approach to stalling) and skip them for now. Currently, they're not
capable of stalling the disk longer 50us (see https://github.com/cockroachdb/cockroach/issues/95886), which makes them
unreliable at exercising stalls.

Add two new roachtests, disk-stall/dmsetup and disk-stall/cgroup that use
dmsetup and cgroup bandwidth restrctions respectively to reliably induce a
write stall for an indefinite duration.

Informs https://github.com/cockroachdb/cockroach/issues/94373.
Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95865)
<!-- Reviewable:end -->

erikgrinaker commented:
Note that the `failover` tests are meant to be benchmarks that measure lease recovery times under various failure modes, not pass/fail tests, as they're [graphed nightly](https://roachperf.crdb.dev/?filter=&view=failover%2Fnon-system%2Fdisk-stall&tab=gce) and we want to track lease recovery performance over time. Any failing tests would probably prevent the data point from being recorded (?). If we want a pass/fail tests, we should probably consider adding a separate roachtest for it?

jbowens commented:
Ack, I'll split it into a separate roachtest.

jbowens committed:
cmd/roachtest: adapt disk-stall detection roachtest

Move the existing disk-stall/* roachtests under disk-stall/fuse/* (for the FUSE
filesystem approach to stalling) and skip them for now. Currently, they're not
capable of stalling the disk longer 50us (see #95886), which makes them
unreliable at exercising stalls.

Add two new roachtests, disk-stall/dmsetup and disk-stall/cgroup that use
dmsetup and cgroup bandwidth restrctions respectively to reliably induce a
write stall for an indefinite duration.

Informs #94373.
Epic: None
Release note: None

nicktrav reviewed (approved):
Very nice! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95865#-:-NMk1Z4w83k4416JTv2d:bh12bql)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker, @renatolabs, @smg260, and @sumeerbhola)



<!-- Sent from Reviewable.io -->

nicktrav reviewed (commented):
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95865)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker, @renatolabs, @smg260, and @sumeerbhola)



<!-- Sent from Reviewable.io -->

nicktrav reviewed (commented):
Reviewed latest patch set. Still <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> :thumbsup: 

> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95865#-:-NMohuDl4BahjpywsDwx:bvg1f43)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker, @renatolabs, @smg260, and @sumeerbhola)



<!-- Sent from Reviewable.io -->

erikgrinaker commented on pkg/cmd/roachtest/tests/disk_stall.go:
```diff
@@ -15,125 +15,441 @@ import (
+	t           test.Test
+	c           cluster.Cluster
+	readOrWrite []string
+	logsToo     bool
```
We should have tests where we stall the log device independently of the data device.

jbowens replied:
Agreed, but it's a little tricky given the volume-level stalling. The FUSE staller makes it a little easier and does implement stalling of the log device independently. @nicktrav will merge his change to fixup and unskip them after this goes in.

erikgrinaker commented on pkg/cmd/roachtest/tests/disk_stall.go:
```diff
@@ -15,125 +15,441 @@ import (
+	c.Run(ctx, c.All(), "mkdir -p logs-external")
+	c.Run(ctx, c.All(), "ln -s /home/ubuntu/logs-external/cockroach.stdout.log logs/cockroach.stdout.log || true")
+	c.Run(ctx, c.All(), "ln -s /home/ubuntu/logs-external/cockroach.stderr.log logs/cockroach.stderr.log || true")
+	c.Run(ctx, c.All(), "ln -s /home/ubuntu/logs-external/roachprod.log logs/roachprod.log || true")
```
Is symlinking specific files thorough enough? We may want to place all of the log files on a separate device that can be stalled entirely, to ensure that all log files and log rotation logic and such is also tested properly.

jbowens replied:
This is actually exempting these log files from the stalling. I'll remove this; it was a change made while we suspected that the reason the process wouldn't terminate was that `cockroach.sh` redirected stdout and stderr to the stalled disk. This turned out to not be the reason, and we've realized there's no way to reliably terminate the process.

erikgrinaker commented on pkg/cmd/roachtest/tests/disk_stall.go:
```diff
@@ -15,125 +15,441 @@ import (
+			return &cgroupDiskStaller{t: t, c: c, readOrWrite: []string{"write", "read"}}
+		},
+		"cgroup/read-write/logs-too=true": func(t test.Test, c cluster.Cluster) diskStaller {
+			return &cgroupDiskStaller{t: t, c: c, readOrWrite: []string{"write", "read"}, logsToo: true}
```
Is it worth testing all combinations of stallers and log/data devices? Or do we feel like a single staller can reliably test all failure modes?

jbowens replied:
> Or do we feel like a single staller can reliably test all failure modes?

Not really, at the moment. Maybe down the line we can consolidate, but we want to be able to exercise exactly the test that a customer performs, regardless of how they perform it.

erikgrinaker reviewed (approved):
Some thoughts on test coverage, but I'm fine with merging this as-is and expanding coverage later.

jbowens reviewed (commented):
I think this is ready now, if you want to take a look at the latest changes.

There's still one bit that I'd like to look at, but it can happen separately: When `n1` fails, the `workload` doesn't shift the connections it had open to `n1` over to `n2` and `n3`. This results in reduced qps for the rest of the test, because it's running with reduced concurrency. I'm pretty sure it's not because the connections are still open to `n1`, given I've verified all sockets are closed on `n1`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95865#-:-NN96l6f8ANYVejbwlPb:biuiykr)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker, @nicktrav, @renatolabs, @smg260, and @sumeerbhola)



<!-- Sent from Reviewable.io -->

jbowens commented:
TFTRs!

bors r=nicktrav,erikgrinaker
